### DISCUSSION TITLE
Maintain resource state

### DISCUSSION THREAD

**Discussion by PSanetra** [dotnet/dotnet-operator-sdk#637] [Ideas & General Discussion]
Hi,

is there a recommended way how to create, delete or patch a resource via the operator or is this out of scope of this framework?

As far as I can see the operator SDK seems to be support reacting to a change in a desired resource state in the cluster via reconciliation, but does not seem to explicitly support syncing a resource status (e.g. via patch) or the existance of a resource (create, delete) to the resource state in the api server. 

Maybe I am missing something, but [according to K8s documentation](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/#custom-controllers) this functionality is one of the core ideas of a Kubernetes controller, right?

> The Kubernetes controller keeps the current state of Kubernetes objects in sync with your declared desired state.

**Comment by buehler** [comment]
Yes, there is :-)

You can create an `IKubernetesClient` for pretty much every resource you want. From there, you can update/delete/create/replace/updatestatus whatever with that resource.

The version that I'm preparing right now has a KubernetesClientFactory that helps with creating all the clients.

If you plan to update the _same_ resource: https://buehler.github.io/dotnet-operator-sdk/src/KubeOps.Operator/README.html#controller

If you don't: use the IKUbernetesClient (or IKubernetes if you want to do it yourself) and create any resource you want. But then I'd suggest you use the "owner reference" meta field to attach it to some other resouce. With the owner reference in place, the newly created object will be garbage collected when the owner is deleted.

**Reply by buehler** [reply]
For example:

```csharp
[EntityRbac(typeof(V1TestEntity), typeof(V1ConfigMap), Verbs = RbacVerb.All)]
public class V1TestEntityController : IEntityController<V1TestEntity>
{
    private readonly IKubernetesClient<V1ConfigMap> _client = KubernetesClientFactory.Create<V1ConfigMap>();

    public async Task ReconcileAsync(V1TestEntity entity)
    {
        if (await _client.GetAsync(entity.Name(), entity.Namespace()) is not null)
        {
            await _client.CreateAsync(
                new V1ConfigMap(metadata: new V1ObjectMeta(name: entity.Name(), namespaceProperty: entity.Namespace()))
                    .Initialize()
                    .WithOwnerReference(entity));
        }
    }
}
```

**Reply by PSanetra** [reply]
Thank you for that example. We have a very similar case like this!

But in another case we want to create and maintain a resource based on observation of some Kubernetes-unrelated state. I am not sure if this is unusual or even considered as bad practice. But in our use case we want to reconcile other k8s resources based on this state and we think therefore we should project that state as resource into Kubernetes. Maybe this use case could be compared to node-resources in K8s, where nodes register themselves on startup and update status information while the node is running.

But I guess for this second use case we need to build some custom solution, where we call the K8s api from outside of the Operator Framework context, right?

**Reply by buehler** [reply]
Hey @PSanetra 
As far as I know, an operator should only reconcile resources (CRDs and base ones), and not based on other magic. So if you want to reconcile some app state, you should create a CRD for it.

It is totally fine to have CRDs without any spec and whatnot properties, only to create other resources for that matter.

**Reply by PSanetra** [reply]
@buehler Alright, I think then we are taking a valid approach :smile:
